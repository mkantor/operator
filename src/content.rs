extern crate handlebars;
use crate::lib::*;
use handlebars::{Context, Handlebars, RenderContext, Renderable};
use serde::Serialize;
use std::io;
use std::path::{Path, PathBuf};
use thiserror::Error;

#[derive(Error, Debug)]
#[error(
  "Failed to parse template{} from content directory path '{}'.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
  .content_directory_path.to_string_lossy()
)]
pub struct RegisteredTemplateParseError {
    source: handlebars::TemplateError,
    content_directory_path: PathBuf,
}

#[derive(Error, Debug)]
#[error(
  "Failed to parse template{}.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
)]
pub struct UnregisteredTemplateParseError {
    source: handlebars::TemplateError,
}

#[derive(Error, Debug)]
pub enum ContentLoadingError {
    #[error(transparent)]
    TemplateParseError(#[from] RegisteredTemplateParseError),

    #[error(
    "Input/output error when loading{} from content directory path '{}'.",
    .name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or(String::from(" content")),
    .content_directory_path.to_string_lossy()
  )]
    IOError {
        source: io::Error,
        content_directory_path: PathBuf,
        name: Option<String>,
    },
}

#[derive(Error, Debug)]
#[error(
  "Rendering failed for template{}.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
)]
pub struct TemplateRenderError {
    #[from]
    source: handlebars::RenderError,
}

pub struct ContentEngine<'a> {
    template_registry: Handlebars<'a>,
}

impl<'a> ContentEngine<'a> {
    pub fn from_content_directory(
        content_directory_path: &'a Path,
    ) -> Result<Self, ContentLoadingError> {
        let template_registry = {
            let mut template_registry = Handlebars::new();
            template_registry
                .register_templates_directory(".hbs", content_directory_path)
                .map_err(|template_render_error| match template_render_error {
                    handlebars::TemplateFileError::TemplateError(source) => {
                        ContentLoadingError::TemplateParseError(RegisteredTemplateParseError {
                            source,
                            content_directory_path: PathBuf::from(content_directory_path),
                        })
                    }
                    handlebars::TemplateFileError::IOError(source, original_name) => {
                        // Handlebars-rust will use an empty string when the error does not
                        // correspond to a specific path.
                        let name = if original_name.is_empty() {
                            None
                        } else {
                            Some(original_name)
                        };
                        ContentLoadingError::IOError {
                            source,
                            content_directory_path: PathBuf::from(content_directory_path),
                            name,
                        }
                    }
                })?;
            template_registry
        };

        Ok(ContentEngine { template_registry })
    }

    pub fn new_content(
        &self,
        handlebars_source: &str,
    ) -> Result<ContentItem, UnregisteredTemplateParseError> {
        let template = handlebars::Template::compile2(handlebars_source, true)
            .map_err(|source| UnregisteredTemplateParseError { source })?;
        Ok(ContentItem::UnregisteredTemplate {
            template_registry: &self.template_registry,
            template,
        })
    }
}

#[derive(Serialize)]
struct GluonRenderData {
    version: GluonVersion,
}

#[derive(Serialize)]
struct RenderData {
    gluon: GluonRenderData,
}

pub enum ContentItem<'a> {
    /// A named template that exists in the registry. This variant must only be
    /// constructed with names that have already been validated against the
    /// registry.
    RegisteredTemplate {
        template_registry: &'a Handlebars<'a>,
        name_in_registry: &'a str,
    },

    /// An anonymous template for on-the-fly rendering.
    UnregisteredTemplate {
        template_registry: &'a Handlebars<'a>,
        template: handlebars::Template,
    },
}

impl Render for ContentItem<'_> {
    type RenderArgs = GluonVersion;
    type Error = TemplateRenderError;

    fn render(&self, gluon_version: GluonVersion) -> Result<String, Self::Error> {
        let render_data = RenderData {
            gluon: GluonRenderData {
                version: gluon_version,
            },
        };

        let rendered_content = match &self {
            ContentItem::RegisteredTemplate {
                template_registry,
                name_in_registry,
            } => template_registry.render(name_in_registry, &render_data)?,
            ContentItem::UnregisteredTemplate {
                template_registry,
                template,
            } => {
                let context = Context::wraps(render_data)?;
                let mut render_context = RenderContext::new(None);
                template.renders(template_registry, &context, &mut render_context)?
            }
        };

        Ok(rendered_content)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_lib::*;

    #[test]
    fn content_engine_can_be_created_from_valid_content_directory() {
        for &path in &CONTENT_DIRECTORY_PATHS_WITH_VALID_CONTENTS {
            assert!(
                ContentEngine::from_content_directory(Path::new(path)).is_ok(),
                "Content engine could not be created from {}",
                path
            );
        }
    }

    #[test]
    fn content_engine_cannot_be_created_from_invalid_content_directory() {
        for &path in &CONTENT_DIRECTORY_PATHS_WITH_INVALID_CONTENTS {
            assert!(
                ContentEngine::from_content_directory(Path::new(path)).is_err(),
                "Content engine was successfully created from {}, but this should have failed",
                path
            );
        }
    }

    #[test]
    fn new_templates_can_be_rendered() {
        let engine = ContentEngine::from_content_directory(
            arbitrary_content_directory_path_with_valid_content(),
        )
        .expect("Content engine could not be created");

        for &(template, expected_output) in &VALID_TEMPLATES {
            let new_content = engine
                .new_content(template)
                .expect("Template could not be parsed");
            let rendered = new_content
                .render(GluonVersion("0.0.0"))
                .expect(&format!("Template rendering failed for `{}`", template,));
            assert_eq!(
                rendered,
                expected_output,
                "Template rendering for `{}` did not produce the expected output (\"{}\"), instead got \"{}\"",
                template,
                expected_output,
                rendered,
            );
        }
    }

    #[test]
    fn new_content_fails_for_invalid_templates() {
        let engine = ContentEngine::from_content_directory(
            arbitrary_content_directory_path_with_valid_content(),
        )
        .expect("Content engine could not be created");

        for &template in &INVALID_TEMPLATES {
            let result = engine.new_content(template);

            assert!(
                result.is_err(),
                "Content was successfully created for invalid template `{}`, but it should have failed",
                template,
            );
        }
    }

    #[test]
    fn new_templates_can_reference_partials_from_content_directory() {
        let content_directory_path = example_path("valid/partials");
        let engine = ContentEngine::from_content_directory(&content_directory_path)
            .expect("Content engine could not be created");

        let template = "this is partial: {{> ab}}";
        let expected_output = "this is partial: a\nb\n\n";

        let new_content = engine
            .new_content(template)
            .expect("Template could not be parsed");
        let rendered = new_content
            .render(GluonVersion("0.0.0"))
            .expect(&format!("Template rendering failed for `{}`", template,));
        assert_eq!(
            rendered,
            expected_output,
            "Template rendering for `{}` did not produce the expected output (\"{}\"), instead got \"{}\"",
            template,
            expected_output,
            rendered,
        );
    }
}
