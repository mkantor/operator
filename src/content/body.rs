//! This module contains various types of steaming HTTP response bodies
//! (reading from files, capturing stdout of a process, etc). All of these
//! types have an impl for Stream<Item=Result<Bytes, StreamError>>.

use super::StreamError;
use bytes::Bytes;
use futures::future::{Future, FutureExt, LocalBoxFuture};
use futures::Stream;
use std::cmp;
use std::fs::File;
use std::io::ErrorKind::Interrupted;
use std::io::{self, Read, Seek};
use std::mem;
use std::pin::Pin;
use std::process::Child;
use std::task::{Context, Poll};

// FIXME: Should not depend on actix from inside the content module.
use actix_web::error::BlockingError;
use actix_web::web;

type ChunkOperation<'a, T> = LocalBoxFuture<'a, Result<T, BlockingError<StreamError>>>;

fn handle_error(error: BlockingError<StreamError>) -> StreamError {
    match error {
        BlockingError::Error(error) => error,
        BlockingError::Canceled => StreamError::Canceled,
    }
}

/// The simplest body: just some bytes that are already in memory. This is a
/// poor excuse for a "stream" (it just dumps everything on the first poll and
/// then ends) but it's handy so we can always work with streams, even at the
/// cost of some efficiency.
pub struct InMemoryBody(pub Bytes);
impl Stream for InMemoryBody {
    type Item = Result<Bytes, StreamError>;

    fn poll_next(mut self: Pin<&mut Self>, _: &mut Context) -> Poll<Option<Self::Item>> {
        let bytes = mem::take(&mut self.0);
        Poll::Ready(if bytes.is_empty() {
            None
        } else {
            Some(Ok(bytes))
        })
    }
}

/// HTTP response body populated by a local file. This was yoinked [from
/// actix-files's `ChunkedReadFile`](https://github.com/actix/actix-web/blob/web-v3.0.0-beta.3/actix-files/src/lib.rs#L58-L117)
/// and only lightly modified.
pub struct FileBody {
    size: u64,
    offset: u64,
    file: Option<File>,
    next: Option<ChunkOperation<'static, (File, Bytes)>>,
    counter: u64,
}
impl FileBody {
    pub fn try_from_file(file: File) -> Result<Self, io::Error> {
        Ok(Self {
            size: file.metadata()?.len(),
            offset: 0,
            file: Some(file),
            next: None,
            counter: 0,
        })
    }
}
impl Stream for FileBody {
    type Item = Result<Bytes, StreamError>;

    fn poll_next(mut self: Pin<&mut Self>, context: &mut Context) -> Poll<Option<Self::Item>> {
        if let Some(ref mut future) = self.next {
            return match Pin::new(future).poll(context) {
                Poll::Ready(Ok((file, bytes))) => {
                    self.next.take();
                    self.file = Some(file);
                    self.offset += bytes.len() as u64;
                    self.counter += bytes.len() as u64;
                    Poll::Ready(Some(Ok(bytes)))
                }
                Poll::Ready(Err(error)) => Poll::Ready(Some(Err(handle_error(error)))),
                Poll::Pending => Poll::Pending,
            };
        }

        let size = self.size;
        let offset = self.offset;
        let counter = self.counter;

        if size == counter {
            Poll::Ready(None)
        } else {
            let mut file = self.file.take().expect("Use after completion");
            self.next = Some(
                web::block(move || {
                    let max_bytes = cmp::min(size.saturating_sub(counter), 65_536);
                    let mut buffer = Vec::with_capacity(max_bytes as usize);
                    file.seek(io::SeekFrom::Start(offset))?;
                    file.by_ref().take(max_bytes).read_to_end(&mut buffer)?;
                    Ok((file, Bytes::from(buffer)))
                })
                .boxed_local(),
            );
            self.poll_next(context)
        }
    }
}

/// HTTP response body populated from the stdout of a running process.
pub struct ProcessBody {
    process: Option<Child>,
    next: Option<ChunkOperation<'static, (Option<Child>, Bytes)>>,
}
impl ProcessBody {
    pub fn new(process: Child) -> Self {
        ProcessBody {
            process: Some(process),
            next: None,
        }
    }
}
impl Stream for ProcessBody {
    type Item = Result<Bytes, StreamError>;

    fn poll_next(mut self: Pin<&mut Self>, context: &mut Context) -> Poll<Option<Self::Item>> {
        if let Some(ref mut future) = self.next {
            return match Pin::new(future).poll(context) {
                Poll::Ready(Ok((process, bytes))) => {
                    self.next.take();
                    self.process = process;
                    Poll::Ready(Some(Ok(bytes)))
                }
                Poll::Ready(Err(e)) => {
                    self.process = None; // Give up on the process after hitting an error.
                    Poll::Ready(Some(Err(handle_error(e))))
                }
                Poll::Pending => Poll::Pending,
            };
        }

        let mut process = match self.process.take() {
            // None means the process has terminated; we're all done!
            None => return Poll::Ready(None),
            Some(process) => process,
        };

        let pid = process.id();
        let next = web::block(move || {
            let mut buffer = [0; 32]; // FIXME: 32 bytes is totally arbitrary.
            match process.stdout {
                None => Err(StreamError::ExecutableOutputCouldNotBeCaptured { pid }),
                Some(ref mut stdout) => {
                    match stdout.read(&mut buffer) {
                        Err(error) if error.kind() == Interrupted => {
                            // If the read was interrupted then it can be tried
                            // again on the next poll. Just emit an empty chunk.
                            Ok((Some(process), Bytes::new()))
                        }
                        Err(fatal_error) => Err(StreamError::from(fatal_error)),
                        Ok(0) => {
                            match process.try_wait()? {
                                None => {
                                    // The process is still running, there was just
                                    // no new output.
                                    Ok((Some(process), Bytes::new()))
                                }
                                Some(exit_status) => {
                                    if !exit_status.success() {
                                        let stderr_contents = {
                                            process.stderr.and_then(|mut stderr| {
                                                let mut error_message = String::new();
                                                match stderr.read_to_string(&mut error_message) {
                                                    Err(_) | Ok(0) => None,
                                                    Ok(_) => Some(error_message),
                                                }
                                            })
                                        };

                                        Err(StreamError::ExecutableExitedWithNonzero {
                                            pid,
                                            stderr_contents,
                                            exit_code: exit_status.code(),
                                        })
                                    } else {
                                        // Successful completion.
                                        Ok((None, Bytes::new()))
                                    }
                                }
                            }
                        }
                        Ok(size) => Ok((Some(process), Bytes::copy_from_slice(&buffer[..size]))),
                    }
                }
            }
        })
        .boxed_local();

        self.next = Some(next);
        self.poll_next(context)
    }
}
