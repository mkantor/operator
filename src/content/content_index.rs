use super::content_engine::CanonicalRoute;
use crate::content_directory::ContentFile;
use serde::Serialize;
use std::collections::HashMap;
use thiserror::Error;

#[derive(Error, Debug)]
#[error("Failed to add route '{}' to index: {}", .failed_route, .message)]
pub struct ContentIndexUpdateError {
    failed_route: String,
    message: String,
}

/// A hierarchial tree mapping out content in the registry. Does not actually
/// contain content items, just their routes.
///
/// For example, given the following content directory:
///
/// ```text
/// content/
///   foo.txt
///   bar.html
///   bar/
///     plugh.md.hbs
///     baz/
///       quux.gif
/// ```
///
/// The content index would be:
///
/// ```yaml
/// foo: foo
/// bar: bar
/// bar/:
///   plugh: bar/plugh
///   baz/:
///     quux: bar/baz/quux
/// ```
#[derive(Clone, Serialize)]
#[serde(untagged)]
pub enum ContentIndex {
    File(CanonicalRoute),
    Directory(ContentIndexEntries),
}

#[derive(Clone, Serialize)]
pub struct ContentIndexEntries(HashMap<String, ContentIndex>);
impl ContentIndexEntries {
    pub fn new() -> Self {
        Self(HashMap::new())
    }

    pub fn try_add<C: AsRef<str>>(
        &mut self,
        canonical_route: C,
    ) -> Result<(), ContentIndexUpdateError> {
        let (dirname_components, basename) = {
            let mut path_components = canonical_route.as_ref().split(ContentFile::PATH_SEPARATOR);
            let basename = path_components.next_back();
            (path_components, basename)
        };

        match basename {
            None => Ok(()), // Successfully inserted nothing! ðŸŽ‰
            Some(basename) => {
                // Navigate to the correct place in the index by iterating path
                // components (except the last one), creating new directories
                // along the way as needed (think mkdir -p).
                let mut node = self;
                for path_component in dirname_components {
                    let next_node = node
                        .0
                        // Non-leaf nodes in the index end with `/` (they
                        // represent directories).
                        .entry(format!("{}/", path_component))
                        .or_insert_with(|| ContentIndex::Directory(Self::new()));

                    node = match next_node {
                        ContentIndex::Directory(branch) => branch,
                        ContentIndex::File(CanonicalRoute(conficting_route)) => {
                            // Each component in dirname_components represents
                            // a directory along the path
                            return Err(ContentIndexUpdateError {
                            failed_route: String::from(canonical_route.as_ref()),
                              message: format!(
                                "There is already a file at '{}', but that needs to be a directory to accommodate the new route.",
                                conficting_route,
                              )
                            });
                        }
                    };
                }

                // Use the last path component to insert a file in the index.
                match node.0.get(basename) {
                    Some(existing_entry) => {
                        let entry_description = match existing_entry {
                            ContentIndex::Directory(..) => "directory",
                            ContentIndex::File(..) => "file",
                        };
                        Err(ContentIndexUpdateError {
                            failed_route: String::from(canonical_route.as_ref()),
                            message: format!(
                                "There is already a {} at '{}'.",
                                entry_description,
                                canonical_route.as_ref(),
                            ),
                        })
                    }
                    None => {
                        node.0.entry(String::from(basename)).or_insert_with(|| {
                            ContentIndex::File(CanonicalRoute::new(canonical_route))
                        });
                        Ok(())
                    }
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn index_has_the_correct_structure() {
        let mut index = ContentIndexEntries::new();
        index.try_add("foo").unwrap();
        index.try_add("bar").unwrap();
        index.try_add("bar/plugh").unwrap();
        index.try_add("bar/baz/quux").unwrap();

        let actual_json = serde_json::to_value(index).unwrap();
        let expected_json = json!({
            "foo": "foo",
            "bar": "bar",
            "bar/": {
              "plugh": "bar/plugh",
              "baz/": {
                "quux": "bar/baz/quux"
              }
            }
        });
        assert_eq!(actual_json, expected_json);
    }
}
