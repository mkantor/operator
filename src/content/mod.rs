mod body;
mod content_directory;
mod content_engine;
mod content_index;
mod content_item;
mod content_registry;
mod handlebars_helpers;
mod mime;
mod route;
mod test_lib;

use crate::bug_message;
use bytes::Bytes;
use content_item::RenderingFailedError;
use futures::Stream;
use serde::Serialize;
use std::io;
use thiserror::Error;

pub use self::mime::{MediaRange, MediaType};
pub use content_directory::ContentDirectory;
pub use content_engine::{
    ContentEngine, ContentLoadingError, FilesystemBasedContentEngine, TemplateError,
};
pub use content_index::ContentIndex;
pub use content_item::UnregisteredTemplate;
pub use content_registry::{ContentRepresentations, RegisteredContent};
pub use route::Route;

// This is just a trait alias to help make type signatures a bit saner.
pub trait ByteStream: Stream<Item = Result<Bytes, StreamError>>
where
    Self: Unpin,
{
}
impl<T> ByteStream for T where T: Stream<Item = Result<Bytes, StreamError>> + Unpin {}

/// A piece of rendered content along with its media type.
pub struct Media<Content: ByteStream> {
    pub media_type: MediaType,
    pub content: Content,
}
impl<Content: ByteStream> Media<Content> {
    fn new(media_type: MediaType, content: Content) -> Self {
        Self {
            media_type,
            content,
        }
    }
}

/// Indicates that it was not possible to produce rendered output, either
/// because rendering was attempted and failed or because no acceptable media
/// types are available.
#[derive(Error, Debug)]
pub enum RenderError {
    #[error(transparent)]
    RenderingFailed(RenderingFailedError),

    #[error("The requested content cannot be rendered as an acceptable media type.")]
    CannotProvideAcceptableMediaType,

    #[doc(hidden)]
    #[error("{} This should never happen: {}", bug_message!(), .0)]
    Bug(String),
}

/// Indicates that something went wrong after starting to stream content.
#[derive(Error, Debug)]
pub enum StreamError {
    #[error(
        "Process exited with {}{}",
        match .exit_code {
            Some(code) => format!("code {}", code),
            None => String::from("unknown code"),
        },
        .stderr_contents.as_ref().map(|message| format!(": {}", message)).unwrap_or_default(),
    )]
    ExecutableExitedWithNonzero {
        pid: u32,
        exit_code: Option<i32>,
        stderr_contents: Option<String>,
    },

    #[error("Executable output could not be captured")]
    ExecutableOutputCouldNotBeCaptured { pid: u32 },

    #[error("Input/output error during rendering")]
    IOError {
        #[from]
        source: io::Error,
    },

    #[error("Stream was cancelled")]
    Canceled,
}

pub trait Render {
    type Output;
    fn render<'engine, 'accept, ServerInfo, QueryParameters, Engine, Accept>(
        &self,
        context: RenderContext<'engine, ServerInfo, QueryParameters, Engine>,
        acceptable_media_ranges: Accept,
    ) -> Result<Media<Self::Output>, RenderError>
    where
        ServerInfo: Clone + Serialize,
        QueryParameters: Clone + Serialize,
        Engine: ContentEngine<ServerInfo>,
        Accept: IntoIterator<Item = &'accept MediaRange>,
        Self::Output: ByteStream;
}

// These must match up with serialized property names in RequestData and
// RenderData.
const TARGET_MEDIA_TYPE_PROPERTY_NAME: &str = "target-media-type";
const REQUEST_DATA_PROPERTY_NAME: &str = "request";
const ROUTE_PROPERTY_NAME: &str = "route";
const QUERY_PARAMETERS_PROPERTY_NAME: &str = "query-parameters";

/// Render data that comes from requests.
#[derive(Clone, Serialize)]
#[serde(rename_all = "kebab-case")]
pub struct RequestData<QueryParameters: Clone + Serialize> {
    /// The request [`Route`] that caused this content to be rendered, if any.
    pub route: Option<Route>,

    /// A parsed version of the request URI's query string.
    pub query_parameters: QueryParameters,
}

/// Data passed to handlebars templates and executables.
///
/// Fields serialize into kebab-case (e.g. `server_info` becomes `server-info`).
#[derive(Clone, Serialize)]
#[serde(rename_all = "kebab-case")]
pub struct RenderData<ServerInfo: Clone + Serialize, QueryParameters: Clone + Serialize> {
    /// A hierarchial index of the content. This is serialized with the name
    /// `/` (with handlebars escaping this looks like `[/].[foo/].bar`).
    #[serde(rename = "/")]
    pub index: ContentIndex,

    /// Metadata about the server, such as its version.
    pub server_info: ServerInfo,

    /// The best [`MediaType`] as determined by content negotiation. Rendering
    /// must emit content in this media type.
    pub target_media_type: Option<MediaType>,

    /// Data that comes from requests.
    pub request: RequestData<QueryParameters>,

    /// An [HTTP `4xx` or `5xx` status code](https://datatracker.ietf.org/doc/html/rfc7231#section-6)
    /// indicating that something went wrong. This will be set while rendering
    /// content for the `--error-handler-route`.
    pub error_code: Option<u16>,
}

/// Values used during rendering, including the data passed to handlebars
/// templates and executables.
pub struct RenderContext<'engine, ServerInfo, QueryParameters, Engine>
where
    ServerInfo: Clone + Serialize,
    QueryParameters: Clone + Serialize,
    Engine: ContentEngine<ServerInfo>,
{
    content_engine: &'engine Engine,
    data: RenderData<ServerInfo, QueryParameters>,
}

impl<'engine, ServerInfo, QueryParameters, Engine>
    RenderContext<'engine, ServerInfo, QueryParameters, Engine>
where
    ServerInfo: Clone + Serialize,
    Engine: ContentEngine<ServerInfo>,
    QueryParameters: Clone + Serialize,
{
    pub fn into_error_context(self, error_code: u16) -> Self {
        RenderContext {
            data: RenderData {
                error_code: Some(error_code),
                ..self.data
            },
            ..self
        }
    }
}
