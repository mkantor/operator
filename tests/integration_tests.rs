use actix_web::client::Client as HttpClient;
use actix_web::http::StatusCode;
use actix_web::test::unused_addr;
use content::ContentDirectory;
use mime_guess::MimeGuess;
use regex::Regex;
use std::collections::hash_map::DefaultHasher;
use std::collections::BTreeMap;
use std::collections::HashMap;
use std::env;
use std::ffi::OsStr;
use std::hash::Hasher;
use std::io::Write;
use std::process::{Command, Stdio};
use std::str;
use std::thread;
use std::time;

// Pull in some utilities from the main crate.
#[path = "../src/content/mod.rs"]
mod content;
#[path = "../src/lib.rs"]
mod lib;
#[path = "../src/test_lib.rs"]
mod test_lib;

use test_lib::*;

fn soliton_command<I, S>(args: I) -> Command
where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    let target_dir = env::current_exe()
        .ok()
        .map(|mut path| {
            path.pop();
            if path.ends_with("deps") {
                path.pop();
            }
            path
        })
        .unwrap();

    let bin_path = target_dir.join(format!("soliton{}", env::consts::EXE_SUFFIX));

    let mut soliton = Command::new(bin_path);
    soliton.args(args);
    soliton
}

/// Attempts to render all non-hidden files in ContentDirectory, returning
/// them as a map of Route -> RenderedContent | ErrorMessage.
fn render_everything(content_directory: ContentDirectory) -> HashMap<String, String> {
    let mut content = HashMap::new();
    for content_file in &content_directory {
        if !content_file.is_hidden() {
            let route = content_file.relative_path_without_extensions();
            let empty_string = String::from("");
            let first_filename_extension =
                content_file.extensions().first().unwrap_or(&empty_string);

            // Target media type is just the source media type.
            let target_media_type = MimeGuess::from_ext(first_filename_extension)
                .first()
                .unwrap_or(mime::STAR_STAR);

            let result =
                render_one_thing(&content_directory, route, &target_media_type.to_string());

            let output_or_error_message = match result {
                Ok(output) => {
                    let hash = {
                        let mut hasher = DefaultHasher::new();
                        hasher.write(&output);
                        hasher.finish()
                    };
                    match String::from_utf8(output) {
                        Ok(string) => string,
                        Err(_) => format!("binary data with hash {:x}", hash),
                    }
                }
                Err(error_bytes) => {
                    String::from_utf8(error_bytes).expect("Error message was not UTF-8")
                }
            };

            content.insert(
                String::from(content_file.relative_path()),
                output_or_error_message,
            );
        }
    }
    content
}

fn render_one_thing(
    content_directory: &ContentDirectory,
    route: &str,
    accept: &str,
) -> Result<Vec<u8>, Vec<u8>> {
    let mut command = soliton_command(&[
        "get",
        &format!(
            "--content-directory={}",
            content_directory
                .root()
                .to_str()
                .expect("Content directory root path was not UTF-8")
        ),
        &format!("--route={}", route),
        &format!("--accept={}", accept),
    ]);
    let output = command.output().expect("Failed to execute process");

    if output.status.success() {
        Ok(output.stdout)
    } else {
        Err(output.stderr)
    }
}

#[test]
fn examples_match_snapshots() {
    for content_directory in example_content_directories() {
        let content_directory_root = &content_directory.root();

        let log_prefix_regex = {
            let datetime_pattern = r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d+[-+]\d{2}:\d{2}";
            let log_level_pattern = r"ERROR|WARN|INFO|DEBUG|TRACE";
            let log_prefix_pattern =
                format!("^({}) - ({}) - ", datetime_pattern, log_level_pattern);
            Regex::new(&log_prefix_pattern).unwrap()
        };

        let unordered_content = render_everything(content_directory);
        let contents = unordered_content
            .iter()
            // If dynamic content files mention where the repo is checked
            // out, redact it to keep tests portable.
            .map(|(key, value)| (key, value.replace(PROJECT_DIRECTORY, "$PROJECT_DIRECTORY")))
            // Also remove the prefix used on log messages.
            .map(|(key, value)| (key, String::from(log_prefix_regex.replace_all(&value, ""))))
            // Files can be omitted from snapshots with a naming convention.
            .filter(|(key, _)| !key.starts_with("SKIP-SNAPSHOT-"))
            .collect::<BTreeMap<_, _>>();

        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_input_file(content_directory_root);
        let id = content_directory_root
            .strip_prefix(example_path("."))
            .unwrap()
            .to_string_lossy();
        insta_settings.set_snapshot_suffix(id);
        insta_settings.bind(|| insta::assert_yaml_snapshot!(contents));
    }
}

#[test]
fn missing_subcommand_is_error() {
    let mut command = soliton_command(&[] as &[&str]);
    let output = command.output().expect("Failed to execute process");

    assert!(
        !output.status.success(),
        "Executing `{:?}` succeeded when it should have failed",
        command
    );
}

#[test]
fn invalid_subcommand_is_error() {
    let mut command = soliton_command(&["invalid-subcommand"]);
    let output = command.output().expect("Failed to execute process");

    assert!(
        !output.status.success(),
        "Executing `{:?}` succeeded when it should have failed",
        command
    );
}

#[test]
fn render_subcommand_succeeds() {
    let input = "{{#if true}}hello world{{/if}}";
    let expected_output = "hello world";

    let mut command = soliton_command(&[
        "render",
        "--content-directory=/dev/null",
        "--media-type=text/plain",
    ]);
    command
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    let mut child = command.spawn().expect("Failed to spawn process");
    child
        .stdin
        .as_mut()
        .expect("Failed to capture child process stdin")
        .write_all(input.as_bytes())
        .expect("Failed to write to child process stdin");
    let output = child.wait_with_output().expect("Failed to execute process");

    assert!(
        output.status.success(),
        "Executing `{:?}` failed when it should have succeeded",
        command
    );
    assert_eq!(
        String::from_utf8(output.stdout).expect("Output was not valid UTF-8"),
        expected_output,
        "Executing `{:?}` did not produce the expected output",
        command
    );
}

#[test]
fn get_subcommand_succeeds() {
    let expected_output = "hello world\n";

    let mut command = soliton_command(&[
        "get",
        &format!(
            "--content-directory={}",
            &example_path("hello-world").to_str().unwrap()
        ),
        "--route=hello",
        "--accept=text/*",
    ]);
    let output = command.output().expect("Failed to execute process");

    assert!(
        output.status.success(),
        "Executing `{:?}` failed when it should have succeeded: {}",
        command,
        String::from_utf8(output.stderr).unwrap_or(String::from(
            "Unable to display error message because stderr was not UTF-8"
        ))
    );
    assert_eq!(
        String::from_utf8(output.stdout).expect("Output was not valid UTF-8"),
        expected_output,
        "Executing `{:?}` did not produce the expected output",
        command
    );
}

#[actix_rt::test]
async fn serve_subcommand_succeeds() {
    let server_address = unused_addr();

    let expected_response_body = "hello world\n";
    let mut command = soliton_command(&[
        "serve",
        &format!(
            "--content-directory={}",
            &example_path("hello-world").to_str().unwrap()
        ),
        &format!("--socket-address={}", server_address),
        "--index-route=hello",
    ]);
    command
        .stdin(Stdio::inherit())
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit());
    let mut child = command.spawn().expect("Failed to spawn process");

    let request = HttpClient::new()
        .get(format!("http://{}/", server_address))
        .header(
            "accept",
            "application/msword, text/*;q=0.9, image/gif;q=0.1",
        );

    // Give the server a chance to start up before sending the request.
    // TODO: Would be better to poll by retrying a few times if the connection
    // is refused.
    thread::sleep(time::Duration::from_millis(500));

    let mut response = request.send().await.expect("Unable to send HTTP request");
    let response_body = response.body().await.expect("Unable to get response body");
    let response_content_type = response
        .headers()
        .get("content-type")
        .expect("Response was missing content-type header");

    assert_eq!(
        response.status(),
        StatusCode::OK,
        "Response status was not 200 OK"
    );
    assert_eq!(
        response_content_type, "text/html",
        "Response content-type was not text/html",
    );
    assert_eq!(
        response_body, expected_response_body,
        "Response body was incorrect"
    );

    child.kill().expect("Failed to kill server");
}
