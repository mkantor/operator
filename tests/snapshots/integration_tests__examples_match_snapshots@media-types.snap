---
source: tests/integration_tests.rs
expression: contents
input_file: src/examples/media-types
---
cannot-become-html.xyz: ""
echo-target-media-type.html.hbs: text/html
html-variation.htm: ""
includes-echo-target-media-type.html.hbs: text/html
nesting/html-that-includes-txt.html.hbs: "Rendering failure: Rendering failed for template: Error rendering \"nesting/html-that-includes-txt\" line 1, col 1: The `get` helper's first argument must be a string (the route of the content item to get), but it was `null`.\nError: Unable to render content.\n\nCaused by:\n    0: Rendering failed for template: Error rendering \"nesting/html-that-includes-txt\" line 1, col 1: The `get` helper's first argument must be a string (the route of the content item to get), but it was `null`.\n    1: Error rendering \"nesting/html-that-includes-txt\" line 1, col 1: The `get` helper's first argument must be a string (the route of the content item to get), but it was `null`.\n"
nesting/just-html.html: "<marquee>wowowow</marquee>\n"
nesting/txt-that-includes-html.txt.hbs: "Rendering failure: Rendering failed for template: Error rendering \"nesting/txt-that-includes-html\" line 1, col 1: The `get` helper's first argument must be a string (the route of the content item to get), but it was `null`.\nError: Unable to render content.\n\nCaused by:\n    0: Rendering failed for template: Error rendering \"nesting/txt-that-includes-html\" line 1, col 1: The `get` helper's first argument must be a string (the route of the content item to get), but it was `null`.\n    1: Error rendering \"nesting/txt-that-includes-html\" line 1, col 1: The `get` helper's first argument must be a string (the route of the content item to get), but it was `null`.\n"
template-cannot-become-html.xyz.hbs: ""
template-html-variation.htm.hbs: ""
